cps/kni.c

```diff
@@ -668,6 +668,231 @@ rd_modroute(const struct nlmsghdr *req, struct route_update *update)
 	return MNL_CB_OK;
 }
 
+static void
+rd_fill_getroute_reply(struct cps_config *cps_conf, struct nlmsghdr *reply,
+	struct gk_fib *fib, int family, uint32_t seq,
+	uint8_t prefix_len, struct ipaddr **gw_addr)
+{
+	struct rtmsg *rm;
+
+	reply->nlmsg_type = RTM_NEWROUTE;
+	reply->nlmsg_flags = NLM_F_MULTI;
+	reply->nlmsg_seq = seq;
+	reply->nlmsg_pid = cps_conf->nl_pid;
+
+	rm = mnl_nlmsg_put_extra_header(reply, sizeof(*rm));
+	rm->rtm_family = family;
+	rm->rtm_dst_len = prefix_len;
+	rm->rtm_src_len = 0;
+	rm->rtm_tos = 0;
+
+	rm->rtm_table = RT_TABLE_MAIN;
+	rm->rtm_scope = RT_SCOPE_UNIVERSE;
+	rm->rtm_type = RTN_UNICAST;
+
+	rm->rtm_flags = 0;
+
+	switch (fib->action) {
+	case GK_FWD_GRANTOR:
+		mnl_attr_put_u32(reply, RTA_OIF, cps_conf->back_kni_index);
+		rm->rtm_protocol = fib->u.grantor.protocol;
+		*gw_addr = &fib->u.grantor.eth_cache->ip_addr;
+		break;
+	case GK_FWD_GATEWAY_FRONT_NET:
+		mnl_attr_put_u32(reply, RTA_OIF, cps_conf->front_kni_index);
+		rm->rtm_protocol = fib->u.gateway.protocol;
+		*gw_addr = &fib->u.gateway.eth_cache->ip_addr;
+		break;
+	case GK_FWD_GATEWAY_BACK_NET:
+		mnl_attr_put_u32(reply, RTA_OIF, cps_conf->back_kni_index);
+		rm->rtm_protocol = fib->u.gateway.protocol;
+		*gw_addr = &fib->u.gateway.eth_cache->ip_addr;
+		break;
+	default:
+		rte_panic("Invalid FIB action (%u) in FIB while being processed by CPS block in %s\n",
+			fib->action, __func__);
+		break;
+	}
+}
+
+static inline void
+rd_send_batch(struct cps_config *cps_conf, struct mnl_nlmsg_batch *batch,
+	uint32_t seq, uint32_t pid)
+{
+	/* Address of routing daemon. */
+	struct sockaddr_nl rd_sa;
+	struct nlmsghdr *done;
+
+	memset(&rd_sa, 0, sizeof(rd_sa));
+	rd_sa.nl_family = AF_NETLINK;
+	rd_sa.nl_groups = 0;
+	rd_sa.nl_pid = pid;
+
+	done = mnl_nlmsg_put_header(mnl_nlmsg_batch_current(batch));
+	done->nlmsg_type = NLMSG_DONE;
+	done->nlmsg_flags = NLM_F_MULTI;
+	done->nlmsg_seq = seq;
+	done->nlmsg_pid = cps_conf->nl_pid;
+
+	if (sendto(mnl_socket_get_fd(cps_conf->rd_nl),
+			mnl_nlmsg_batch_head(batch),
+			mnl_nlmsg_batch_size(batch) + sizeof(*done),
+			0, (struct sockaddr *)&rd_sa, sizeof(rd_sa)) < 0) {
+		CPS_LOG(ERR,
+			"sendto: cannot dump route batch to daemon: %s\n",
```

Comment:
Need to specify the daemon: ``sendto: cannot dump route batch to routing daemon``